include/pybind11/cast.h

```diff
@@ -2203,6 +2203,18 @@ object object_api<Derived>::call(Args &&...args) const {
 
 PYBIND11_NAMESPACE_END(detail)
 
+
+template<typename T>
+type type::of() {
+   static_assert(
```

Comment:
nice idea to check

Reply:
I'd change the error message though: ``py::type::of<T> only supports the case where T is a regis[..]`` instead of the generic "This"